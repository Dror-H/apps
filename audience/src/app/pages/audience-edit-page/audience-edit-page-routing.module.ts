import { UserHasSubscriptionGuard } from '@audience-app/auth/guards/user-has-subscription/user-has-subscription.guard';
import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import { AudienceEditPageComponent } from '@audience-app/pages/audience-edit-page/audience-edit-page.component';
import { PendingChangesGuard } from '@instigo-app/ui/shared';

const routes: Routes = [
  {
    path: '',
    component: AudienceEditPageComponent,
    canActivate: [UserHasSubscriptionGuard],
    canDeactivate: [PendingChangesGuard],
  },
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule],
})
export class AudienceEditPageRoutingModule {}
