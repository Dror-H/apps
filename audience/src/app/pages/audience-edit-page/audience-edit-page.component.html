<ng-container ingoPendingChanges>
  <div class="divider--xl"></div>
  <nz-affix class="brand-affix">
    <div class="brand-container" nz-row nzAlign="middle" nzJustify="center">
      <a nz-col routerLink="audience-search">
        <img alt="instigo logo" class="brand-image" src="assets/svg/instigo-brand.svg" />
      </a>
    </div>
  </nz-affix>
  <div (click)="toggleOverview(false)" [ngClass]="{ inview: isOverviewActive }" class="sticky-overview-overlay"></div>
  <form [formGroup]="audienceForm" class="ingo-add-edit-form">
    <nz-card>
      <div [nzGutter]="24" nz-row>
        <div [nzSpan]="10" nz-col>
          <nz-form-item>
            <nz-form-label>Audience Name</nz-form-label>
            <nz-form-control>
              <input formControlName="name" nz-input placeholder="Audience Name" />
              <span
                *ngIf="audienceForm.controls.name.errors && !audienceForm.controls.name.pristine"
                class="text-danger mt-5 d-block"
              >
                Min. length of 4 letters is required
              </span>
            </nz-form-control>
          </nz-form-item>
        </div>
      </div>
      <ingo-facebook-saved-audience-targeting
        [audienceForm]="audienceForm"
        [rules]="rules"
      ></ingo-facebook-saved-audience-targeting>
      <ingo-select-tags [config]="selectTagsConfig" formControlName="userTags"></ingo-select-tags>
      <div class="divider--sm"></div>
      <div nz-row nzJustify="end">
        <button
          (click)="goBack()"
          class="margin-right btn-outline btn-light"
          nz-button
          nz-col
          nzType="default"
          type="button"
        >
          <span>{{ isAudienceSaved ? 'Search New Audience' : 'Back' }}</span>
        </button>
        <div *ngIf="user?.isAdmin" class="tooltip-pos margin-right">
          <ingo-explain
            *ngIf="audienceForm.controls.reach.errors?.incorrectReach"
            [tooltipType]="'danger'"
            tooltipId="app.audience.segmentError"
          ></ingo-explain>
          <button
            (click)="triggerSaveAudience()"
            [disabled]="isAudienceSaved || audienceForm.status === 'INVALID'"
            [ngClass]="{ 'save-button--saved': isAudienceSaved }"
            [nzLoading]="isSavingAudience$ | async"
            nz-button
            nz-col
            nzType="primary"
          >
            <span>{{ isAudienceSaved ? 'Audience Saved' : 'Save Audience' }}</span>
          </button>
        </div>

        <div class="tooltip-pos">
          <ingo-explain
            *ngIf="audienceForm.controls.reach.errors?.incorrectReach"
            [tooltipType]="'danger'"
            tooltipId="app.audience.segmentError"
          ></ingo-explain>
          <button
            (click)="triggerExportAudience()"
            [disabled]="(isAudienceExported && isAudienceSaved) || audienceForm.status === 'INVALID'"
            [ngClass]="{ 'save-button--saved': isAudienceExported && isAudienceSaved }"
            [nzLoading]="isSavingAudience$ | async"
            nz-button
            nz-col
            nzType="primary"
          >
            <span>{{
              isAudienceExported && isAudienceSaved ? 'Audience Exported' : 'Export To Facebook Ads Manager'
            }}</span>
          </button>
        </div>
      </div>
    </nz-card>
    <div
      (click)="toggleOverview(true)"
      [ngClass]="{ inview: isOverviewActive }"
      [nzXl]="6"
      class="sticky-overview sticky-audience-summary"
      nz-col
      nzSpan="8"
    >
      <ingo-audience-summary
        [sticky]="true"
        [audienceForm]="audienceForm"
        [isOverviewActive]="isOverviewActive"
      ></ingo-audience-summary>
    </div>
  </form>
</ng-container>
