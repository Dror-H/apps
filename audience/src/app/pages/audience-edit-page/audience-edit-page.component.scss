@use 'backgrounds.variables' as backgrounds;
@use 'colors.variables' as colors;
@use 'components/header.variables' as header;

.ingo-add-edit-form {
  display: flex;
  gap: 1rem;

  nz-card {
    flex-basis: 100%;
  }

  ingo-facebook-saved-audience-targeting {
    flex: 1;
  }

  .margin-right {
    margin-right: 0.5rem;
  }

  .save-button--saved {
    background-color: colors.$green;
  }
}

.brand {
  &-container {
    height: header.$header-height;
    left: 50%;
    position: relative;
  }

  &-image {
    width: 115px;
  }

  &-affix {
    max-width: 50%;
    position: relative;
    z-index: 22;
  }
}

.tooltip-pos {
  position: relative;

  ingo-explain {
    font-size: 16px;
    position: absolute;
    right: -8px;
    top: -9px;
    z-index: 200;
  }
}
