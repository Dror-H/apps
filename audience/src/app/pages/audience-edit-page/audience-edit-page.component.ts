import { Location } from '@angular/common';
import { Component, Inject, OnDestroy, OnInit } from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { ActivatedRoute, Router } from '@angular/router';
import { User } from '@audience-app/global/models/app.models';
import { AudiencesService } from '@audience-app/global/services/audiences/audiences.service';
import { LoadingService } from '@audience-app/global/services/loading/loading.service';
import { ModalService } from '@audience-app/global/services/modal/modal.service';
import {
  createFacebookExportUrl,
  createRulesFromTargeting,
  getAudiencesFlatValues,
} from '@audience-app/pages/audience-edit-page/utils';
import { AudiencesState } from '@audience-app/store/audiences.state';
import { LoadingState } from '@audience-app/store/loading.state';
import { UserState } from '@audience-app/store/user.state';
import {
  AdAccountDTO,
  AdSetDTO,
  AudienceType,
  CampaignDTO,
  SearchResult,
  SupportedProviders,
  TargetingDto,
  UserTargetings,
} from '@instigo-app/data-transfer-object';
import { getProcessedValues, SelectTagsConfig, STORE } from '@instigo-app/ui/shared';
import { SelectSnapshot } from '@ngxs-labs/select-snapshot';
import { Select, Store } from '@ngxs/store';
import { BehaviorSubject, MonoTypeOperatorFunction, Observable } from 'rxjs';
import { debounceTime, retry, skip, skipWhile, switchMap, tap } from 'rxjs/operators';
import { SubSink } from 'subsink';

@Component({
  selector: 'audi-audience-edit-page',
  templateUrl: './audience-edit-page.component.html',
  styleUrls: ['./audience-edit-page.component.scss'],
})
export class AudienceEditPageComponent implements OnInit, OnDestroy {
  @SelectSnapshot(AudiencesState.getSelectedAudiences) public selectedAudiences: SearchResult[];
  @Select(LoadingState.getIsSavingAudience) public isSavingAudience$: Observable<boolean>;

  public rules: TargetingDto;
  public audienceForm = new FormGroup({});
  public isOverviewActive = false;
  public saveAudience$ = new BehaviorSubject<AdAccountDTO>(null);
  public isAudienceSaved = false;
  public isAudienceExported = false;
  public selectTagsConfig: SelectTagsConfig;
  public user: User;

  private subscriptions = new SubSink();

  constructor(
    @Inject(STORE) private store: Store,
    private fb: FormBuilder,
    private _location: Location,
    private _router: Router,
    private audiencesService: AudiencesService,
    private route: ActivatedRoute,
    private modalService: ModalService,
    private loadingService: LoadingService,
  ) {}

  ngOnInit(): void {
    this.initAudienceForm();
    this.subscribeToChangeAdAccount();
    this.subscribeToSaveAudience();
    this.enableSaveOnValueChange();
    this.handleMergedAudienceFromRouteData();
  }

  ngOnDestroy(): void {
    this.subscriptions.unsubscribe();
  }

  public goBack(): void {
    if (this.isAudienceSaved) {
      void this._router.navigateByUrl('/');
      return;
    }

    const locationState = this._location.getState();
    const isFirstAppNavigation = (locationState as { navigationId: number }).navigationId === 1;
    isFirstAppNavigation ? void this._router.navigateByUrl('/') : this._location.back();
  }

  public toggleOverview(toggleState: boolean): void {
    this.isOverviewActive = toggleState;
  }

  public triggerSaveAudience(): void {
    this.saveAudience$.next(null);
  }

  public triggerExportAudience(): void {
    this.modalService.openAdAccountSelectorModal(this.rules).afterClose.subscribe((adAccount) => {
      if (adAccount) {
        this.saveAudience$.next(adAccount);
      }
    });
  }

  private subscribeToSaveAudience(): void {
    this.subscriptions.add(
      this.saveAudience$
        .pipe(
          skip(1),
          debounceTime(300),
          skipWhile(() => this.isAudienceSaved),
          switchMap((value) => {
            this.loadingService.isSavingAudience.emit(true);
            if (!value) {
              return this.saveAudience();
            }
            return this.exportAudience$(value);
          }),
          this.cleanupAfterSaveAudience(),
          retry(), //* prevent observable complete
        )
        .subscribe(),
    );
  }

  private cleanupAfterSaveAudience(): MonoTypeOperatorFunction<unknown> {
    return tap({
      next: () => {
        this.isAudienceSaved = true;
        this.audiencesService.resetSelectedAudiences();
        this.audiencesService.setFoundAudiences.emit(null);
        this.loadingService.isSavingAudience.emit(false);
      },
      error: () => {
        this.loadingService.isSavingAudience.emit(false);
      },
    });
  }

  private exportAudience$(value: AdAccountDTO): Observable<{ campaign: CampaignDTO; adSet: Partial<AdSetDTO> }> {
    return this.exportAudience(value).pipe(
      tap((result) => {
        this.modalService.openOnOK(
          createFacebookExportUrl(value.id, result.campaign.providerId, result.adSet.providerId),
        );
      }),
    );
  }

  private handleMergedAudienceFromRouteData(): void {
    const mergedAudience: SearchResult[] = this.route.snapshot.data[0];
    this.handleMergedAudience(mergedAudience);
  }

  private handleMergedAudience(mergedAudienceArr: SearchResult[]): void {
    this.setAudienceRules(mergedAudienceArr[0]?.spec as TargetingDto);
    this.setUserTags(mergedAudienceArr);
    this.setSelectTagsConfig();
  }

  private setAudienceRules(mergedAudience: TargetingDto): void {
    this.rules = createRulesFromTargeting(mergedAudience);
  }

  private setUserTags(selectedAudiences: SearchResult[]): void {
    const tags = getProcessedValues(getAudiencesFlatValues(selectedAudiences, 'userTags'), 65);
    this.audienceForm.get('userTags').setValue(tags);
  }

  private enableSaveOnValueChange(): void {
    this.subscriptions.add(
      this.audienceForm.valueChanges
        .pipe(
          tap((val) => {
            this.isAudienceSaved = false;
            this.isAudienceExported = false;
          }),
        )
        .subscribe(),
    );
  }

  private setSelectTagsConfig(): void {
    this.selectTagsConfig = {
      disabled: false,
      tagSuggestions: getProcessedValues(getAudiencesFlatValues(this.selectedAudiences, 'topics'), 150),
    };
  }

  private saveAudience(): Observable<UserTargetings> {
    return this.audiencesService.saveAudience(this.getAudienceValues());
  }

  private exportAudience(adAccount: AdAccountDTO): Observable<{ campaign: CampaignDTO; adSet: Partial<AdSetDTO> }> {
    return this.audiencesService.exportAudience({ ...this.getAudienceValues(), adAccount }).pipe(
      tap(() => {
        this.isAudienceExported = true;
      }),
    );
  }

  private getAudienceValues(): {
    name: string;
    targeting: TargetingDto;
    userTags: string[];
  } {
    const provider = this.audienceForm.get('provider').value;
    const name = this.audienceForm.get('name').value;
    const { include, exclude } = this.audienceForm.get('target').value as TargetingDto;
    const targeting: TargetingDto = { provider, include, exclude };
    const userTags: string[] = this.audienceForm.get('userTags').value;
    return { name, targeting, userTags };
  }

  private initAudienceForm(): void {
    this.audienceForm = this.fb.group({
      audienceType: [AudienceType.SAVED_AUDIENCE],
      audienceSubType: [null, []],
      adAccount: ['', [Validators.required]],
      provider: [SupportedProviders.FACEBOOK],
      description: [''],
      name: ['', [Validators.minLength(4), Validators.pattern(/^(\s+\S+\s*)*(?!\s).*$/)]],
      reach: ['', [Validators.required]],
      target: [],
      userTags: [],
    });
  }

  private subscribeToChangeAdAccount(): void {
    this.subscriptions.add(
      this.store.select(UserState).subscribe((user) => {
        if (user?.adAccounts?.length) {
          this.audienceForm.get('adAccount').setValue({ providerId: user.adAccounts[0].id });
          this.user = user;
        }
      }),
    );
  }
}
