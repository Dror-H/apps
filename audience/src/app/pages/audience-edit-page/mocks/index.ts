export * from './inits';
export * from './search-results.mocks';
