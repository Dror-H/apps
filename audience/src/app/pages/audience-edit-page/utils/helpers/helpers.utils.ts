import { SearchResult } from '@instigo-app/data-transfer-object';
import { flatten } from 'lodash';

export function getAudiencesFlatValues(audiences: SearchResult[], key: 'topics' | 'userTags'): string[] {
  if (!audiences?.length) {
    return [];
  }
  return flatten(audiences.filter((a) => !!a[key]).map((audience) => audience[key]));
}
