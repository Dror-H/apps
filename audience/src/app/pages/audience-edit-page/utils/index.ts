export * from './audience-rules/audience-rules.utils';
export * from './helpers/helpers.utils';
export * from './facebook-export-url/facebook-export-url';
