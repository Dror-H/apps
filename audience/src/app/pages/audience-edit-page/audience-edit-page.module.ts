import { NzButtonModule } from 'ng-zorro-antd/button';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { AudienceApiService } from '@audience-app/api/audience-api/audience-api.service';
import { TargetingApiService } from '@audience-app/api/targeting-api/targeting-api.service';
import { AudienceEditPageRoutingModule } from '@audience-app/pages/audience-edit-page/audience-edit-page-routing.module';
import { UiFacebookModule } from '@instigo-app/ui/facebook';
import {
  AUDIENCES_API_SERVICE,
  ENVIRONMENT,
  PendingChangesGuard,
  STORE,
  TARGETING_API_SERVICE,
  UiSharedModule,
} from '@instigo-app/ui/shared';
import { TranslateModule } from '@ngx-translate/core';
import { NgxsModule, Store } from '@ngxs/store';
import { NzCardModule } from 'ng-zorro-antd/card';
import { NzFormModule } from 'ng-zorro-antd/form';
import { NzGridModule } from 'ng-zorro-antd/grid';
import { NzInputModule } from 'ng-zorro-antd/input';
import { environment } from 'src/environments/environment';
import { AudienceEditPageComponent } from './audience-edit-page.component';
import { NzAffixModule } from 'ng-zorro-antd/affix';
import { AudiencesState } from '@audience-app/store/audiences.state';

@NgModule({
  declarations: [AudienceEditPageComponent],
  imports: [
    CommonModule,
    ReactiveFormsModule,
    AudienceEditPageRoutingModule,
    UiFacebookModule,
    UiSharedModule,
    NzGridModule,
    NzFormModule,
    NzInputModule,
    NzCardModule,
    NzButtonModule,
    NzAffixModule,
    NgxsModule.forFeature([AudiencesState]),
    TranslateModule,
  ],
  providers: [
    { provide: PendingChangesGuard, useClass: PendingChangesGuard },
    { provide: STORE, useExisting: Store },
    { provide: TARGETING_API_SERVICE, useClass: TargetingApiService },
    { provide: AUDIENCES_API_SERVICE, useExisting: AudienceApiService },
    { provide: ENVIRONMENT, useValue: environment },
  ],
})
export class AudienceEditPageModule {}
