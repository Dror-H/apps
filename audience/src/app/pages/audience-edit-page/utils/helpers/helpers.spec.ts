import { mockSearchResults } from '@audience-app/pages/audience-edit-page/mocks';
import { getAudiencesFlatValues } from '@audience-app/pages/audience-edit-page/utils/helpers/helpers.utils';

describe('audiences helpers utils', () => {
  describe('getAudiencesFlatValues', () => {
    it('should return empty array', () => {
      const topics = getAudiencesFlatValues([], 'topics');
      const userTags = getAudiencesFlatValues([], 'userTags');
      const undefinedResult = getAudiencesFlatValues(undefined, 'userTags');
      expect(topics).toEqual([]);
      expect(userTags).toEqual([]);
      expect(undefinedResult).toEqual([]);
    });

    it('should return flattened values', () => {
      const topics = getAudiencesFlatValues(mockSearchResults, 'topics');
      const userTags = getAudiencesFlatValues(mockSearchResults, 'userTags');
      expect(topics).toMatchSnapshot();
      expect(userTags).toMatchSnapshot();
    });
  });
});
