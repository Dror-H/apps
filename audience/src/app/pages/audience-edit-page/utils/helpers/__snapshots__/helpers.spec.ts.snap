// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`audiences helpers utils getAudiencesFlatValues should return flattened values 1`] = `
Array [
  "business",
  "director",
  "owners",
  "marketing",
  "managing",
  "topic 1",
  "topic 2",
  "topic 3",
  "topic 4",
  "topic 5",
  "topic 6",
]
`;

exports[`audiences helpers utils getAudiencesFlatValues should return flattened values 2`] = `
Array [
  "gambling",
  "football betting",
  "soccer betting",
  "Engaged Shoppers on Facebooks",
  "NBA fans",
  "motor racing",
  "horse racing",
  "tag 1",
  "tag 2",
  "tag 3",
  "tag 4",
  "tag 5",
  "tag 6",
]
`;
